Source/Particles/Collision/BackgroundMCCCollision.cpp

```diff
@@ -0,0 +1,385 @@
+                                      amrex::Real vp = sqrt(
+                                                            2.0 / mass1 * PhysConst::q_e
+                                                            * (E_coll - scattering_process.m_energy_penalty)
+                                                            );
+                                      RandomizeVelocity(ux[ip], uy[ip], uz[ip], vp, engine);
+                                  }
+                                  break;
+                              }
+                          }
+                          );
+    amrex::Gpu::synchronize();
+}
+
+
+/** Perform MCC ionization interactions
+ *
+ * @param pti particle iterator
+ * @param species1/2 reference to species container used to inject new
+ particles from ionization events
+ *
+ */
+void BackgroundMCCCollision::doBackgroundIonization
+( int lev, WarpXParticleContainer& species1,
+  WarpXParticleContainer& species2)
+{
+    WARPX_PROFILE("BackgroundMCCCollision::doBackgroundIonization()");
+
+    SmartCopyFactory copy_factory_elec(species1, species1);
+    SmartCopyFactory copy_factory_ion(species1, species2);
+    const auto CopyElec = copy_factory_elec.getSmartCopy();
+    const auto CopyIon = copy_factory_ion.getSmartCopy();
+
+    const auto Filter = ImpactIonizationFilterFunc(
+                                                   *m_ionization_processes[0],
+                                                   m_mass1, m_total_collision_prob_ioniz,
+                                                   m_nu_max_ioniz / m_background_density
+                                                   );
+
+    amrex::Real vel_std = std::sqrt(
+                                    PhysConst::kb * m_background_temperature / m_background_mass
+                                    );
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+    for (WarpXParIter pti(species1, lev); pti.isValid(); ++pti) {
+        auto& elec_tile = species1.ParticlesAt(lev, pti);
+        auto& ion_tile = species2.ParticlesAt(lev, pti);
+
+        const auto np_elec = elec_tile.numParticles();
+        const auto np_ion = ion_tile.numParticles();
+
+        auto Transform = ImpactIonizationTransformFunc(
+                                                       m_ionization_processes[0]->m_energy_penalty, m_mass1, vel_std
+                                                       );
+
+        const auto num_added = filterCopyTransformParticles<1>(
+                                                               elec_tile, ion_tile, elec_tile, np_elec, np_ion,
+                                                               Filter, CopyElec, CopyIon, Transform
+                                                               );
+
+        setNewParticleIDs(elec_tile, np_elec, num_added);
+        setNewParticleIDs(ion_tile, np_ion, num_added);
+        amrex::Gpu::synchronize();
```

Comment:
I added this sync call and the one in `doBackgroundCollisionsWithinTile` to get the code to run stably on CUDA. It may be over-synchronizing, especially with this in the loop. Could someone who knows the GPU usage well comment on whether it should be safe to pull it out of the loop?

Reply:
Or even to a single call after the tile or level loops in `doCollisions`?

Reply:
There is already a sync at the end of the `filterCopyTransform` call. The next two function calls, to `SetNewParticleIDs`, only modify local data for the tile currently being operated on. Then there is a synchronize at the end of the `ParIter` loop. So I think this one is not needed, since we don't need the ids on previously processed tiles to be set already inside this loop to do the next iteration.